Sources/NIOFileSystem/FileChunks.swift

```diff
@@ -302,45 +347,75 @@ private struct ProducerState: Sendable {
                 )
                 return .failure(error)
             }
+        case .pausedProducing:
+            return .success(nil)
         case .done:
             return .success(nil)
         }
     }
 
-    mutating func readBytes(_ count: Int) {
+    mutating func didReadBytes(_ count: Int) {
         switch self.state {
         case var .producing(state):
-            if let currentRange = state.range {
-                let newRange = (currentRange.lowerBound + Int64(count))..<currentRange.upperBound
-                if newRange.lowerBound >= newRange.upperBound {
-                    assert(newRange.lowerBound == newRange.upperBound)
-                    self.state = .done
-                } else {
-                    state.range = newRange
-                    self.state = .producing(state)
-                }
+            if state.updateRange(count: count) {
+                self.state = .done(emptyRange: false)
             } else {
-                if count == 0 {
-                    self.state = .done
-                }
+                self.state = .producing(state)
+            }
+        case var .pausedProducing(state):
+            if state.updateRange(count: count) {
+                self.state = .done(emptyRange: false)
+            } else {
+                self.state = .pausedProducing(state)
             }
         case .done:
             ()
         }
     }
 
-    mutating func readEnd() {
+    mutating func didReadEnd() {
         switch self.state {
-        case .producing:
-            self.state = .done
+        case .pausedProducing, .producing:
+            self.state = .done(emptyRange: false)
         case .done:
             ()
         }
     }
 
+    mutating func pauseProducing() {
+        switch self.state {
+        case .producing(let state):
+            self.state = .pausedProducing(state)
+        case .pausedProducing, .done:
+            ()
+        }
+    }
+
     mutating func done() {
-        self.state = .done
+        self.state = .done(emptyRange: false)
     }
 }
 
+
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+extension ProducerState.Producing {
+    mutating func updateRange(count: Int) -> Bool {
```

Comment:
It's not obvious _how_ the range is updated without looking at the implementation, also the return value is quite misleading: `if updateRange(count: blah) { ... }` looks a whole lot like "if the range was updated then ...".

Would probably be clearer if you call this `didReadBytes(_ count: Int)` and documented the return value (you could use an enum for the return type but that's probably overkill)